apps/emqx_retainer/src/emqx_retainer_index.erl

```diff
@@ -0,0 +1,144 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2020-2022 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_retainer_index).
+
+-export([foreach_index_key/3,
+         convert_tokens_for_index/2,
+         index_score/2,
+         select_index/2,
+         condition/1,
+         condition/2,
+         restore_topic/3
+        ]).
+
+-export_type([index/0]).
+
+-type index() :: list(pos_integer()).
+
+-type index_key() :: {index(), {emqx_topic:words(), emqx_topic:words()}}.
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+-spec foreach_index_key(fun((index_key()) -> any()), list(index()), emqx_topic:words()) -> ok.
+foreach_index_key(_Fun, [], _Tokens) -> ok;
+
+foreach_index_key(Fun, [Index | Indexes], Tokens) ->
+    Key = {Index, convert_tokens_for_index(Index, Tokens)},
+    Fun(Key),
+    foreach_index_key(Fun, Indexes, Tokens).
+
+-spec convert_tokens_for_index(index(), emqx_topic:words()) ->
+    {emqx_topic:words(), emqx_topic:words()}.
+convert_tokens_for_index(Index, Tokens) ->
+    split_index_tokens(Index, Tokens, 1, [], []).
+
+-spec index_score(index(), emqx_topic:words()) -> non_neg_integer().
+index_score(Index, Tokens) ->
+    index_score(Index, Tokens, 1, 0).
+
+-spec select_index(emqx:words(), list(index())) -> index() | undefined.
+select_index(Tokens, Indexes) ->
+    select_index(Tokens, Indexes, 0, undefined).
+
+-spec condition(index(), emqx_topic:words()) -> ets:match_pattern().
+condition(Index, Tokens) ->
+    condition(Index, Tokens, 1, [], []).
+
+-spec condition(emqx_topic:words()) -> ets:match_pattern().
+condition(Tokens) ->
+    Tokens1 = [case W =:= '+' of true -> '_'; _ -> W end || W <- Tokens],
+    case length(Tokens1) > 0 andalso lists:last(Tokens1) =:= '#' of
+        false -> Tokens1;
+        _ -> (Tokens1 -- ['#']) ++ '_'
```

Comment:
This code looks wrong:
1) It creates an improper list (terminated with '_' instead of []). Is it the intention?
2) https://www.erlang.org/doc/man/ets.html#type-match_pattern type is a `tuple() | atom()`, but this function can return a list.

Reply:
1. Yes, `[a, b, c | '_']` spec matches `[a, b, c | _ ]` lists.
2. Yes, type is wrong. The function returns part of match head.

Reply:
One side note: if topics are not validated before this function is called, an user could pass in something (invalid) such as `#/a/#`, and only the first `#` will be removed, instead of the last one.